### ISSUE TITLE
Test failures in connect

### ISSUE THREAD

**Issue by dfandrich** [wojtekka/libgadu#32]
The test suite produces 84 test failures on an x86_64 Linux system with (mostly) the latest dependencies (e.g., gcc 9.3.0, glibc 2.31, gnutls 3.6.12) with libgadu 1.12.2. You can see the failure [in this log file](http://pkgsubmit.mageia.org/autobuild/cauldron/aarch64/core/2020-03-19/libgadu-1.12.2-2.mga8.src.rpm/build.0.20200319152711.log) but without details. The first failure shows up like this;

```
325/648: sync,  80 open,    8074 open,    443 open,    resolver open,    server no,  proxy no,  ssl yes
** gg_login(0x7fff0a675490: [uin=1, async=0, ...]);
** gg_watch_fd(0xee1040);
// gg_watch_fd() GG_STATE_RESOLVE_HUB_SYNC
// gg_watch_fd() GG_STATE_CONNECT_HUB
// gg_watch_fd() connecting to 127.0.0.1:80
** gg_connect(127.0.0.1, 80, 0);
// gg_watch_fd() GG_STATE_CONNECTING_HUB
// gg_watch_fd() GG_STATE_SEND_HUB
// sending http query:
GET /appsvc/appmsg_ver11.asp?tls=1&fmnumber=1&fmt=2&lastmsg=0&version=11.3.45.10771&age=2&gender=1 HTTP/1.0
Connection: close
Host: appmsg.gadu-gadu.pl

// gg_watch_fd() GG_STATE_READING_HUB
// gg_watch_fd() GG_STATE_READING_HUB
// received http reply:
HTTP/1.0 200 OK

0 0 127.0.0.1:443 127.0.0.1
reply=0, host="127.0.0.1:443"
// gg_watch_fd() GG_STATE_RESOLVE_GG_SYNC
// gg_watch_fd() GG_STATE_CONNECT_GG
resolver_index=0, connect_index=0, connect_port={443,0}
// gg_watch_fd() connecting to 127.0.0.1:443
** gg_connect(127.0.0.1, 443, 0);
// gg_watch_fd() GG_STATE_CONNECTING_GG
// gg_watch_fd() connected
Handshake failed: -50, The request is invalid.
 // gg_session_init_ssl() gnutls_set_default_priority failed
// gg_session_connect() could not login
** gg_event_free(0xecd220);
** gg_free_session(0xee1040);
```

I'm not familiar with the package so I can't debug this effectively.

**Comment by wojtekka** [comment]
You mention "x86_64" architecture, yet the the log contains "aarch64" target, so I'm a bit confused. I looked at previous aarch64 build results and the oldest one from 2020-01-20 with gnutls 3.6.11 also fails. Some older x86_64 build succeeds with gnutls 3.6.8.

With limited set of build logs available it's hard to tell whether it's an issue with "aarch64" architecture or recent gnutls. I tried building libgadu on x86_64 Fedora with gnutls 3.6.12 and saw no issues. I wasn't able to get aarch64 Fedora running. 

So a couple of questions:
1. Is it possible to see a _recent_ x86_64 build log (i.e. with gnutls 3.6.12) so check if it's aarch64 specific?
2. Is it possible to get a temporary account on a aarch64 Mageia system to perform some manual tests?
3. Are you aware about any Mageia-specific modifications to gnutls package, e.g. hardened default configuration, disabled legacy ciphers, key exchange protocols etc.?

**Comment by dfandrich** [comment]
It's failing with aarch64, too. I guess the build bot stops after the aarch64 failure and doesn't bother with x86_64. You can see the x86_64 build/test log and connect.log at https://gist.github.com/dfandrich/c5249a974e3bdd716665446b952e40b7  You can see the Mageia GnuTLS configuration at https://svnweb.mageia.org/packages/cauldron/gnutls/current/SPECS/gnutls.spec?revision=1552778&view=markup  which doesn't have any patches. The CFLAGS are in the log, namely CFLAGS='-O2 -g -pipe -Wformat -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fstack-protector --param=ssp-buffer-size=4 -fasynchronous-unwind-tables'

**Comment by wojtekka** [comment]
It took me a while, but I finally managed to install Mageia Cauldron and build libgadu there. I do not see this issue though. gcc 10.1.0, glibc 2.31, gnutls 3.6.13. Does it still occur on build servers? If yes, is there any way to get temporary access to a system where I could look into it?

**Comment by dfandrich** [comment]
On Tue, May 12, 2020 at 01:40:52AM -0700, Wojtek Kaniewski wrote:
> It took me a while, but I finally managed to install Mageia Cauldron and build
> libgadu there. I do not see this issue though. gcc 10.1.0, glibc 2.31, gnutls
> 3.6.13. Does it still occur on build servers? If yes, is there any way to get
> temporary access to a system where I could look into it?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.*
> 

That's odd. I kicked off another build on the build system, but it's still
failing:
http://pkgsubmit.mageia.org/uploads/failure/cauldron/core/release/20200512084225.danf.duvel.19103/log/libgadu-1.12.2-3.mga8/build.0.20200512084254.log
The build systems are locked down so access there isn't possible, but I'm able
to reproduce it locally in a Mageia Cauldron chroot with a simple:

  tar xavf ~/rpm/SOURCES/libgadu-1.12.2.tar.gz
  cd libgadu-1.12.2/
  ./configure
  make -j9
  make -j9 check

Compiling that way in the host Mageia 7 system works fine. I also tried a
bastard compile on mga7 using the Cauldron gnutls binaries, and the tests
passed there, too.

That was puzzling, so I grepped libgnutls to look for any config files that it
might be using. Sure enough, it uses
/etc/crypto-policies/back-ends/gnutls.config which didn't exist in the chroot.
Since the logs show the error "Handshake failed: -50, The request is invalid."
it's conceivable that a configuration issue it to blame.  I copied that file
into the chroot and lo and behold, the tests passed!

So, it seems the default crypto policies used by libgnutls aren't sufficient
for the tests to pass. Installing a full Cauldron system would install the
policies due to them getting pulled in through another dependency, but the
minimal chroot image that both I and the build system are using don't pull them
in. I took a quick look as to why and they're listed as a dependency so it's
weird why they're not getting pulled in.

In any case, it looks like it's not a libgadu problem. Thanks for taking the
time to help get to the bottom of this!
